src/main/cms/cms.c

```diff
@@ -553,13 +553,13 @@ static void cmsDrawMenu(displayPort_t *pDisplay, uint32_t currentTimeUs)
     uint32_t room = displayTxBytesFree(pDisplay);
 
     if (pDisplay->cleared) {
-        for (p = pageTop, i= 0; p->type != OME_END; p++, i++) {
+        for (p = pageTop, i= 0; p->type != OME_END && i < CMS_MAX_ROWS; p++, i++) {
             SET_PRINTLABEL(runtimeEntryFlags[i]);
             SET_PRINTVALUE(runtimeEntryFlags[i]);
         }
         pDisplay->cleared = false;
     } else if (drawPolled) {
-        for (p = pageTop, i = 0; p <= pageTop + pageMaxRow ; p++, i++) {
+        for (p = pageTop, i = 0; (p <= pageTop + pageMaxRow) && i < CMS_MAX_ROWS ; p++, i++) {
```

Comment:
I think this needs the `p->type != OME_END && ...` condition too.

Reply:
actually, since `pageMaxRow` is calculated using OME_END you could perhaps use `p <= pageTop + pageMaxRow` instead of the `i < CMS_MAX_ROWS` in all of these.

`pageMaxRow` could also be limited by CMS_MAX_ROWS too, in one place.

Either way, a consistent way of checking to see if the `p` value is within range is preferred.

Reply:
So we're clear, these loops all iterate from the first item and to the last to be drawn, they all have one source of data but right now there are three different ways to see if the loop should stop.  `cmsUpdateMaxRow` is already doing the calculation of pageMaxRow so likely the loops should all have the same identical condition of `(p <= pageTop + pageMaxRow)` only.

Reply:
@hydra Thanks for your input.  So you're saying use `(p <= pageTop + pageMaxRow)` and also limit `pageMaxRow `with `CMS_MAX_ROWS `where it is calculated, i.e. in one place?

Reply:
yes, i think that will work but my theory needs testing.

Reply:
Tested it - it does work.